### ISSUE TITLE
User Generate Packs - Pack Opening Simulator

### ISSUE THREAD

**Issue by AlanOC91** [AlanOC91/YGOPRODeck#198]
This was a request that is potentially doable but would require significant time/investment to get right.

The basic request is to allow a user to **create/upload a custom Pack** that they can then open/simulate. 

I'm thinking the workflow would be like so:

- Create a Pack Page with various inputs
- Fill out pack details: Pack Name, Number of Cards (drop-down box - 3, 5 , 9 for start), Custom Image?
- Pack Contents: This is difficult, possibly allow user to upload CSV with exact card names and rarities? If card name not exact, card is skipped. These would have to go into some sort of custom database table along with the user who uploaded them and pack name. A database join is then needed to get full card detail.  Will also need a limit on how many cards can be in a pack. I'm thinking 100 for the beginning. 
- Prices won't be included since it wouldn't make sense.
- Once info has been filled out, user hits upload and data is stored.
- User can then visit a custom pack opener page (it will be kept separate from the official pack opener page) where they can open the custom packs they created. They should also have the option to delete the pack.
- If a user needs to share the custom pack, they could possibly just give someone the CSV file and that user can create the custom pack themselves? Not ideal but no great solution here. Unless maybe we give a share link that a log in user can add the pack to their account?

I like the idea in general but I'm concerned about the amount of work necessary and how popular it may even be. If more users request it or if the idea holds some weight then I may start work on it.

**Comment by f-master** [comment]
I was planning to ask, how can people make their own packs? Do you have a place where people can download the pack simulator code and change the archive to their own. I am willing to set up the OCG sets.

**Comment by AlanOC91** [comment]
> I was planning to ask, how can people make their own packs? Do you have a place where people can download the pack simulator code and change the archive to their own. I am willing to set up the OCG sets.

Not quite how it would work. It would have to auto generate ratios depending on custom set card rarities that the user sets. So it would then pull a generic common -> super rare -> ultra rare -> secret rare algorithm and use that for each custom pack. 

This is only conceptual at the moment. Having users set pre-defined pack opening coding is way beyond the scope of this.

**Comment by AexDEye9** [comment]
i'm not the greatest fan of csv, it's not a widely used file extension and it feels wierd overall.
i don't really know any alternatives either, though... so maybe a custom file format should be made? i prefer puretext

**Comment by AlanOC91** [comment]
I was going to close this but I considered approaching it from a different POV:

I have just released a very much work in progress feature for "Custom Packs" in the pack opening simulator. Currently this works through the collection manager "box" feature. If you have a "box" created in your collection manager with over 20 cards in it, you can open this box as a custom pack (and share the url with others to allow them to open it too). It's very much an early stage feature and I'm in the phase where I need some feedback to potentially improve it. You should see a new button in your collection manager that will allow you to open the box as a pack.

![Open Box as Pack](https://user-images.githubusercontent.com/37552458/103633501-b698d200-4f3d-11eb-9302-779404ccac39.png)

![CustomPack](https://user-images.githubusercontent.com/37552458/103633506-b7c9ff00-4f3d-11eb-88b1-637126e281cf.png)

![CustomPackPull](https://user-images.githubusercontent.com/37552458/103633517-bac4ef80-4f3d-11eb-94c7-f9fc5869fb2e.png)

**Comment by forsetite** [comment]
Just curious - would it be possible to add a draft mode on this as well?

**Comment by AlanOC91** [comment]
> Just curious - would it be possible to add a draft mode on this as well?

Yep it should be! I can add it to the mountain of TODOs I have. But definitely doable!

**Comment by KyrieDevles** [comment]
I would love for this to continue development. I hope the ability to create a fully custom pack (cards and everything) would be implemented soon, as well as draft mode.

**Comment by 4meUnkle** [comment]
> 
> 
> > Just curious - would it be possible to add a draft mode on this as well?
> 
> Yep it should be! I can add it to the mountain of TODOs I have. But definitely doable!

Is there any chance the pack opener could have an Open All Packs at once button if you're doing multiple packs?

**Comment by threwnie** [comment]
Just noticed this - looks awesome; do you need help getting any of the requested stuff set up? I agree that a draft mode would be awesome.

**Comment by AexDEye9** [comment]
> > Given the new updates to the pack opener, will there be an update to custom packs any time soon? Just curious!
> 
> Yep! I have plans for a brand new system for custom packs. It's still in the super super early stages so still a decent way off. The general idea is that you get to a "deck builder" type page and add cards you want into a custom pack and assign them a custom rarity. Then you can save the pack and make it publicly available.

sounds amazing! i would LOVE if it were possible to use some sort of code, or pseudo scripting language to write out everything from scratch, from custom rarities with custom colors, to even custom pools of cards like with the dark beginning sets, it'd love to see it!

**Comment by Richarizardd** [comment]
Hi @AlanOC91 - chiming in to say that this would be a great feature. Your Pack Simulator is very useful for doing custom drafts in DuelingNexus. Having the ability to create a custom set would be very useful for testing Cube formats!

**Comment by lovelydumpling** [comment]
I logged in for the first time for a long time (took a pretty big break from YuGiOh after burnout) and noticed the feature's been implemented! That's really awesome.

**Comment by TheUntamed** [comment]
Hey everyone. I would like to suggest an extansion to this feature.
Something I'm missing on all the card managing tools and web applications out there. And that's the ability to handle ### custom cards.
With platforms like YGO Omega being able to create and use custom cards all I need now is a way to handle my custom cards, create custom packs with them and maybe share them with others.

Now, to elaborate the idea in some more detail:
Just like there is already a csv upload for custom packs containing 2 columns (card name and rarity) for custom cards it would have to contain the other card details as well. The same you can already filter for in the personal card collection (card type, level, etc.)
Regarding the card images. I can understand if that might become a problem if people upload too many cards. So maybe by default cards can be uploaded without image and get a placeholder image when shown in image view and adding an image is for premium users only. I personally would be fine with that and even pay for the ability to do pack drafts or maybe even cube drafts including custom cards.

**Comment by AlanOC91** [comment]
@TheUntamed Hello! 

This is potentially doable. As you said, uploading a CSV with the full card information could in theory allow this. However, it's quite a tall ask to handle this on the blackened as we would then need to store all of this information.

Card Images are another story and I think for security and safety reasons (any user could in theory upload any type of image) we would want to avoid something like that.

Let me mull on it for a bit.

**Comment by AlanOC91** [comment]
@TheUntamed Hello!

We just had a team meeting regarding your request. It's a good idea, but it'd require tons of changes all over the place since our systems aren't built for it, so we're not looking to pursue it right now.
